Fix mutating Paint bug

justinmc opened the pull request:
A tricky bug was encountered in https://github.com/flutter/flutter/pull/170321/files#r2138619243 where mutating Paint caused [paints](https://main-api.flutter.dev/flutter/flutter_test/paints.html) to behave strangely.

Originally this PR just added something to the docs for `paints` to give people a heads up, but now it fixes the root cause. Yak shave :razor: .

Fixes https://github.com/flutter/flutter/issues/171194

justinmc committed:
Add a warning about mutated Paint instances to paints

chunhtai commented on packages/flutter_test/lib/src/mock_canvas.dart:
```diff
@@ -46,6 +46,10 @@ import 'test_async_utils.dart';
 ///
+/// This matcher does not work well when [Paint] instances are mutated and
+/// reused. Create a new instance of [Paint] for subsequent draw calls instead
+/// of mutating and reusing one instance.
```
reusing the paint is fine, but it is just our TestRecordingCanvas can't handle it. I think we should fix the TestRecordingCanvas instead.

justinmc replied:
Good point, that should be doable. I'll take a look.

justinmc committed:
Merge branch 'master' into mutable-paint

justinmc committed:
Actually, remove the docs change and fix the root cause.

This only fixes it for drawCircle. I'll have to implement all the other
canvas calls.

justinmc commented:
This will require adding similar changes for all of the other methods in Canvas. Moving to draft for now.

justinmc committed:
Other Canvas methods that use Paint

justinmc committed:
Test for each method

justinmc committed:
Use the normal createTestImage

ValentinVignal commented on packages/flutter_test/test/recording_canvas.dart:
Doesn't it be renamed `recording_canvas_test.dart` in order to be run in the test pipelines?

justinmc replied:
Beginner mistake, thanks.

justinmc committed:
_test

justinmc committed:
Fix incorrect CupertinoActivityIndicator test

justinmc committed:
Merge branch 'master' into mutable-paint

justinmc committed:
Also do same CupertinoProgressIndicatorFix for adaptive progress indicator

justinmc committed:
Weird formatting again

justinmc committed:
Merge branch 'master' into mutable-paint
